code/obj/machinery/manufacturer.dm

```diff
@@ -1026,44 +1026,47 @@ TYPEINFO(/obj/machinery/manufacturer)
 		storage[index_2] = temp_hold
 		src.should_update_static = TRUE
 
-	proc/eject_material(var/mat_ref, mob/user = null)
-		if (src.mode != MODE_READY)
-			src.grump_message(user, "You cannot eject materials while the unit is working.")
-			return
-		var/ejectamt = 0
-		var/turf/ejectturf = get_output_location(user)
-		var/obj/item/material_piece/target
-		for(var/obj/item/material_piece/P in src.get_contents())
-			if ("\ref[P]" == mat_ref)
-				target = P
-				break
-		if (isnull(target))
-			src.grump_message(user, "ERROR: Material not found in storage.", sound = TRUE)
+	proc/eject_material(var/mat_ref, mob/user)
+		var/obj/item/material_piece/P = src.get_material_by_ref(mat_ref)
+		if (!src.can_eject_material(P, user))
 			return
-		if (target.amount < 1)
-			src.grump_message(user, "ERROR: Not enough material to eject bars.", sound = TRUE)
+		var/eject_amount = tgui_input_number(user,
+											 "How many material pieces do you want to eject?",
+											 title = "Eject Materials",
+											 default = 0,
+											 max_value = ceil(P.amount),
+											 min_value = 0,
+											 round_input = TRUE,
+											)
+		if (!src.can_eject_material(P, user))
 			return
-		// Since tgui_input_number() is awaited here, make sure they are adjacent before and after
-		if (!src.check_physical_proximity(user))
+		// In case tgui_input_number() misses something or it's 0 (Likely trying to cancel action)
+		if (!isnum(eject_amount) || !isnum_safe(eject_amount) || eject_amount < 1)
+			src.grump_message(user, "ERROR: Cannot eject [eject_amount] bars.", sound = TRUE)
 			return
-		ejectamt = floor(tgui_input_number(user,"How many material pieces do you want to eject?","Eject Materials", 0, ceil(target.amount), 0))
-		if (!src.check_physical_proximity(user) || src.mode != MODE_READY || ejectamt <= 0 || !isnum_safe(ejectamt))
-			return
-		if (!ejectturf)
-			return
-		if (ejectamt > target.amount)
-			src.grump_message(user, "There's not that much material in [name]. It has ejected what it could.", sound = TRUE)
-			ejectamt = floor(target.amount)
-		if (ejectamt == target.amount)
-			if (user)
-				user.put_in_hand_or_drop(target)
-			src.storage.transfer_stored_item(target, ejectturf)
+		// This should never happen either
+		if (eject_amount > P.amount)
+			src.grump_message(user, "ERROR: Cannot eject [eject_amount] bars, as there are only [floor(P.amount)] bars available to eject.", sound = TRUE)
```

Comment:
Some materials (e.g. cotton) aren't bars. "pieces", maybe? Or is there a proc to determine the correct noun somewhere?

Reply:
Ah yeah, gonna have to find a proc for that. Not sure if there is one really, I may just have to avoid referring to the quantity of bars/sheets/blahblahblah of fabric